WIP: New Achivement and Badge: Joined a Guild

CooperFLe opened the pull request:
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # (Put Issue # or URL here, if applicable. This will automatically close the issue if your PR is merged in)
Working on #8614
### Changes
[//]: # (Describe the changes that were made in detail here. Include pictures if necessary)

So far, I have the badge working, but the user must log out and back in the see the badge after earning it. 

It was suggested that I need to sync the client user with the server, but I don't think that is the case because of where I added the code. The party on and party up badges are in the same place and work just fine. 

I'm hoping posting this PR will help someone help me figure out my problem. I'm probably just completely overlooking something obvious. Sorry to ask for help when I'm supposed to be the one helping out here!

[//]: # (Put User ID in here - found in Settings -> API)

----
UUID:

Commit:
Add Joined a Guild achivement

Added achievement information and sprite

Commit:
Add Joined a Guild achivement

Added achievement information and sprite

Commit:
Add Joined a Guild achivement

Added achievement information and sprite

Commit:
Added check to award Joined a Guild achivement

Bug: Only shows achivement after user logs out and logs back in after earning it.

Commit:
Added check to award Joined a Guild achivement

Bug: Only shows achivement after user logs out and logs back in after earning it.

Commit:
Added check to award Joined a Guild achivement

Bug: Only shows achivement after user logs out and logs back in after earning it.

SabreCat committed:
3.85.0

paglias commented on test/api/v3/integration/groups/POST-groups_invite.test.js:
```diff
@@ -4,9 +4,11 @@ import {
   translate as t,
 } from '../../../../helpers/api-v3-integration.helper';
 import { v4 as generateUUID } from 'uuid';
+import nconf from 'nconf';
```
I think you pulled some changes from other PRs

CooperFLe replied:
That does appear to have happened... How do I undo that...

SabreCat committed:
New User Tasks for Mobile (#8682)

* feat(mobile): different default tasks

* fix(linting): missing space

* fix(user): correct client type logic

* test(integration): tasks by platform

* fix(test): remove only

* test(user): deeper checks on tasks

* refactor(test): whitespace for readability

SabreCat committed:
feat(subs): Jackalope Pets (#8684)

SabreCat committed:
chore(sprites): compile

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
3.86.0

paglias commented:
Hmm. You may try to switch to another branch, cherry pick (git cherry-pick
commitNumber) the commits you need and then create a new PR from the new
branch

Il 18 Apr 2017 20:31, "Cooper F Le" <notifications@github.com> ha scritto:

> *@CooperFLe* commented on this pull request.
> ------------------------------
>
> In test/api/v3/integration/groups/POST-groups_invite.test.js
> <https://github.com/HabitRPG/habitica/pull/8667#discussion_r112030355>:
>
> > @@ -4,9 +4,11 @@ import {
>    translate as t,
>  } from '../../../../helpers/api-v3-integration.helper';
>  import { v4 as generateUUID } from 'uuid';
> +import nconf from 'nconf';
>
> That does appear to have happened... How do I undo that...
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/HabitRPG/habitica/pull/8667#discussion_r112030355>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAkAV0rKZQGFERJC-BgepNkFN6LLsUHrks5rxQGVgaJpZM4M7YH2>
> .
>

Commit:
Fix Lint test

Added space after if and before opening if bracket

Commit:
Fix Lint test

Added space after if and before opening if bracket

Commit:
Fix Lint test

Added space after if and before opening if bracket

CooperFLe commented:
@paglias looks like only my changes are part of this PR now.

TheHollidayInn commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -561,6 +561,12 @@ api.joinGroup = {
       }
     }
 
+    if (group.type === 'guild') {
```
Can you add a tests around this new functionality?

CooperFLe replied:
I don't actually know how to write tests. Or even where to start for that. If you can point me in the right direction, I'd be happy to give it a shot though.

TheHollidayInn replied:
You will want to read through these examples: https://github.com/HabitRPG/habitica/blob/develop/test/api/v3/integration/groups/POST-groups_groupId_join.test.js

To run test, use `gulp test:api-v3:integration` from the command line.

It is helpful to note that you can run individual tests by adding `it.only`

TheHollidayInn commented on website/server/controllers/api-v3/groups.js:
```diff
@@ -561,6 +561,12 @@ api.joinGroup = {
 
+    if (group.type === 'guild') {
+      promises.push(User.update({
+        'achievements.guild': {$ne: true},
```
This also will update all users. You will want to limit this to users from this guild only. Actually, I think what you want to do here is update the joining user only. 

So, if (!user. achievements.guild) user. achievements.guild = true or something

SabreCat committed:
feat(content): Fairy Potions, Mistiflying Quests

SabreCat committed:
fix(quest): strings and fetch items

SabreCat committed:
chore(sprites): compile

SabreCat committed:
chore(i18n): update locales

SabreCat committed:
chore(news): Bailey

SabreCat committed:
3.87.0

SabreCat committed:
Merge branch 'release' into develop

Commit:
Merge remote-tracking branch 'origin/develop' into develop

Mathwhiz committed:
Change icon for NPC achievement (#8628)

* Change icon for NPC achievement

* Add npc icon

TheHollidayInn committed:
Added wintery skins to content (#8645)

* Added wintery skins to content

* Removed features json

* Fixed some name issues

Hus274 committed:
Creating api for gifting Hall Of Heroes Item Reference (#8118)

* Creating specific api to vend item reference for gifting items to a user through the hall of heroes. closes #8100

* removed 'When in doubt, ask Tyler' from the hall of heroes line

* Moving hall of heroes item path call from server side to client side

* updating let to var

* Addressing PR comments

Alys committed:
add mountText for all drop eggs and quest eggs (#8678)

I've checked that all these added strings do exist in locales/en/content.json

Commit:
Made uneditable habits show counter reset frequency (#8675)

Added counter reset frequency to counter tooltip
Solves issue https://github.com/HabitRPG/habitica/issues/8571

TheHollidayInn committed:
Added block when user types a swear word listed in banned words (#8197)

* Added block when user types a swear word listed in banned words

* Moved banned words check to server

* Removed unused code

* Moved banned words to separate file and fixed grammar.

* Updated chat test

* Changed error to BadRequest

* Fixed regex matching

* Updated test banned word

* Moved banned words and cached regex

* Updated banned word message

* Add ban filter only for tavern

* Added tavern id constant

* Added more tests for banned words

* Added warning to banned words

* Added alert

* Added new regex to capture markdown

* Fixed lint, spelling and importing

TheHollidayInn committed:
Prevented ios and android subs from being upgrade to group plans (#8646)

* Prevented ios and android subs from being upgrade to group plans

* Add admin email sender

* Updated admin email

* Updated get user email

* Fixed grammar issue

TheHollidayInn committed:
Group managers (#8591)

* Added abiltiy to add group managers

* Added ability to remove managers

* Added ability for managers to add group tasks

* Allower managers to assign tasks

* Allowed managers to unassign tasks

* Allow managers to delete group tasks

* Allowed managers to approve

* Added initial ui

* Added approval view for managers

* Allowed managers to edit

* Fixed lint issues

* Added spacing to buttons

* Removed leader from selection of group managers

* Code review updates

* Ensured approvals are only done once

* Added ability for parties to add managers

* Add notifications to all managers when approval is requests

* Removed tasks need approval notifications from all managers when task is approve

* Fixed linting issues

* Hid add managers UI from groups that are not subscribed

* Removed let from front end

* Fixed issues with post task url params

* Fixed string locales

* Removed extra limited strings

* Added cannotedit tasks function

* Added limit fields and notification check by taskId

* Localized string and other minor issues

* Added manager and leader indicator

* Added group notifications refresh on sync

* Added close button for group notifications

* Removed group approval notifications when manager is removed

* Moved leader/manager indicators to after hp

* Added manager fields to groups

* Spelling and syntax fixes

Commit:
Merge remote-tracking branch 'origin/develop' into develop

CooperFLe commented:
Alright, this has gotten out of hand. I am not sure where all of these addition commits are coming from, and every time I try to get rid of them more appear. I'm going to close this PR, and get a new one going later hopefully containing ONLY my changes.
